ADR/ADR047-compute-submission-status-from-timestamps.md

```diff
@@ -0,0 +1,52 @@
+# ADR047: Compute submission status from delivery timestamps
+
+Date: 2025-12-04
+
+## Status
+
+Proposed
+
+## Context
+
+We are making submission delivery asynchronous for all submission methods (email and S3) and we need a more robust way to track the delivery status of submissions in forms-runner.
+
+### Using the Submission model for all submission types
+
+To make S3 submissions asynchronous, we need to use the `Submission` model for S3 submissions. Currently, only email submissions create `Submissions`. The model only supports `pending` or `bounced` values for the `delivery_status` enum - there is no `delivered` status. This would make it difficult for S3 submission to use the `Submission` model, as we would need a way see which Submission are successful versus those that may need to be retried.
+
+### Handling race conditions with async notifications
+
+For email submissions, delivery and bounce notifications are processed asynchronously. This creates a race condition: we may receive and process these notifications out of order. Using timestamps for `delivered_at` and `bounced_at` allows us to accurately determine the sequence of events. We can only mark a submission as `delivered` if we can confirm that any bounce notification didn't occur after the successful delivery notification (i.e. an async bounce).
+
+## Decision
+
+We will replace the fixed `delivery_status` enum attribute with a computed `status` method that derives the submission status from `delivered_at` and `bounced_at` timestamp columns.
+
+The implementation will proceed in three phases:
+
+1. **Add new timestamp column**: Add the `bounced_at` attribute to the `Submission` model via a database migration (the `delivered_at` column already exists).
+
+2. **Switch to computed status**: Replace usage of `delivery_status` with the computed `status` method throughout the codebase. The `status` method will return one of three values:
+   - `:pending` - when both `delivered_at` and `bounced_at` are nil
+   - `:delivered` - when `delivered_at` is present and either `bounced_at` is nil or `delivered_at` is more recent
```

Comment:
Does this line up with our understanding of the timestamps associated with bounce and delivery SNS notifications when we get both? Is the timestamp on the bounce notification always later than the timestamp on the delivered notification?